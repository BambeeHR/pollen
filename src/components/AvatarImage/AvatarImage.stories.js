import { boolean, number, text } from '@storybook/addon-knobs';
import AvatarImage from './AvatarImage.vue';

export default {
  title: 'AvatarImage',
  component: AvatarImage,
  parameters: {
    componentSubtitle: 'Avatar iamge',
  },
};

const image = 'https://cdn.filestackcontent.com/enCa5BTVeIsLHcgkdpAc';

export const Default = () => ({
  components: { AvatarImage },
  data() {
    return { image };
  },
  template: '<AvatarImage :image="image" />',
});

export const NoImage = () => ({
  components: { AvatarImage },
  template: '<AvatarImage />',
});

export const WithInitials = () => ({
  components: { AvatarImage },
  template: '<AvatarImage initials="XO" />',
});

export const WithKnobs = () => {
  return {
    components: { AvatarImage },
    props: {
      border: {
        default: boolean('Border', false),
      },
      image: {
        default: text('Image', image),
      },
      initials: {
        default: text('Initials', 'XO'),
      },
      name: {
        default: text('Name', ''),
      },
      size: {
        default: number('Size', 40),
      },
      square: {
        default: boolean('Square', false),
      },
    },
    template: `
      <AvatarImage
        :border="border"
        :image="image"
        :initials="initials"
        :name="name"
        :size="size"
        :square="square"
      />
    `,
  };
};

/* eslint-disable no-param-reassign */
[Default, NoImage, WithInitials, WithKnobs].forEach((item) => {
  item.story = {
    parameters: {
      jest: ['BaseButton.test.js'],
    },
  };
});
