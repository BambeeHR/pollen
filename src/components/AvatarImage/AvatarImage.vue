<template functional>
  <div
    :class="[
      {
        'avatar-image--rounded': !props.square,
        'avatar-image--bordered': props.border,
      },
      data.class,
      data.staticClass,
    ]"
    class="avatar-image"
    :style="$options.getStyles(props)"
    v-bind="$options.getAttributes({ data, parent })"
    v-on="listeners"
  >
    <div
      v-if="(props.initials || props.name) && !props.image"
      class="avatar-image__initials"
    >
      {{ props.name ? $options.getNameInitials(props.name) : props.initials }}
    </div>
  </div>
</template>

<script>
import DEFAULT_USER_IMAGE from '../../assets/defaultUserImage';
import getAttributes from '../../util/getAttributes';
import getNameInitials from '../../util/getNameInitials';
import resizeFilestackImage from '../../util/resizeFilestackImage';

const pxToRem = (val) => `${parseInt(val, 10) / 16}rem`;

const getStyles = (props) => {
  const { size, image, initials } = props;
  let backgroundImage = null;
  if (image) {
    backgroundImage = `url(${resizeFilestackImage(image, {
      width: size,
      height: size,
    })})`;
  } else if (!initials) {
    backgroundImage = `url(${DEFAULT_USER_IMAGE})`;
  }
  return {
    backgroundImage,
    fontSize: pxToRem(props.size * 0.44),
    height: pxToRem(props.size),
    width: pxToRem(props.size),
  };
};

/**
 * Display a user's image. If that `image` is coming from Filestack, appropriate
 * resize parameters will be applied to match the `size` prop. If no `image` url
 * is supplied, the user's `initials` will display. If both initials and image
 * are missing, a default user image will be displayed.
 */
export default {
  getAttributes,
  getNameInitials,
  getStyles,
  props: {
    /** If true, adds a border around the image. */
    border: {
      type: Boolean,
      default: false,
    },
    /** The url of the image to display. */
    image: {
      type: String,
      default: null,
    },
    /** A fallback pair of initials to display if no image is supplied. */
    initials: {
      type: String,
      default: null,
    },
    /** A full name that will be displayed as initials. */
    name: {
      type: String,
      default: null,
    },
    /** The height/width of the avatar in pixels. */
    size: {
      type: Number,
      default: 40,
    },
    /** If true, displays the image as a square. */
    square: {
      type: Boolean,
      default: false,
    },
  },
};
</script>

<style scoped>
.avatar-image {
  @apply bg-cover 
    bg-center 
    bg-gray-5 
    bg-no-repeat 
    box-border
    flex 
    items-center 
    justify-center;
}

.avatar-image--bordered {
  @apply border border-gray-5 border-solid;
}

.avatar-image--rounded {
  @apply rounded-full;
}

.avatar-image__initials {
  @apply font-semibold text-white tracking-wide;
}
</style>
