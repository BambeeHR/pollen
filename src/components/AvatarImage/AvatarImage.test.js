import { mount } from '@vue/test-utils';
import DEFAULT_USER_IMAGE from '../../assets/defaultUserImage';
import AvatarImage from './AvatarImage.vue';

describe('AvatarImage', () => {
  test('renders correctly', () => {
    const wrapper = mount(AvatarImage, {
      propsData: {
        image: 'https://cdn.filestackcontent.com/enCa5BTVeIsLHcgkdpAc',
        initials: 'XO',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('applies filestack resize parameters', () => {
    const wrapper = mount(AvatarImage, {
      propsData: {
        image: 'https://cdn.filestackcontent.com/enCa5BTVeIsLHcgkdpAc',
        initials: 'XO',
      },
    });
    expect(wrapper.attributes().style).toContain('/resize=width:40,height:40/');
  });

  test('displays intials if there is no image', () => {
    const wrapper = mount(AvatarImage, {
      propsData: {
        initials: 'XO',
      },
    });
    expect(wrapper.text()).toBe('XO');
  });

  test('displays default user image if no initials or image are provided', () => {
    const wrapper = mount(AvatarImage, {});
    expect(wrapper.attributes().style).toContain(DEFAULT_USER_IMAGE);
  });
});
