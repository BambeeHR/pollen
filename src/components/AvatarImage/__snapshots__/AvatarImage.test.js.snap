// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarImage renders correctly 1`] = `
<div
  class="avatar-image avatar-image--rounded"
  style="background-image: url(https://cdn.filestackcontent.com/resize=width:40,height:40/enCa5BTVeIsLHcgkdpAc); font-size: 1.0625rem; height: 2.5rem; width: 2.5rem;"
>
  <!---->
</div>
`;
